File: src/jp/mcbe/fuyutsuki/Texter/text/FloatingTextCluster.php
<?php

declare(strict_types=1);

namespace jp\mcbe\fuyutsuki\Texter\text;

use jp\mcbe\fuyutsuki\Texter\data\Data;
use pocketmine\level\Level;
use pocketmine\math\Vector3;
use JsonSerializable;
use pocketmine\Player;

/**
 * Class FloatingTextCluster
 * @package jp\mcbe\fuyutsuki\Texter\text
 */
class FloatingTextCluster implements Sendable, JsonSerializable {

	use Nameable {
		Nameable::__construct as nameableConstruct;
	}

	/** @var Vector3 */
	private $position;
	/** @var Vector3 */
	private $spacing;
	/** @var FloatingText[] */
	private $floatingTexts = [];

	public function __construct(
		Vector3 $position,
		string $name,
		?Vector3 $spacing = null,
		array $texts = []
	) {
		$this->setPosition($position);
		$this->nameableConstruct($name);
		$this->setSpacing($spacing);
		$this->generateFloatingText($texts);
	}

	public function position(): Vector3 {
		return $this->position;
	}

	public function setPosition(Vector3 $position) {
		$this->position = $position;
	}

	public function spacing(): Vector3 {
		return $this->spacing;
	}

	public function setSpacing(?Vector3 $spacing = null) {
		$this->spacing = $spacing ?? new Vector3;
	}

	public function calculateSpacing(int $index): Vector3 {
		return $this->position->add($this->spacing->multiply($index));
	}

	public function recalculatePosition() {
		foreach ($this->floatingTexts as $index => $floatingText) {
			$floatingText->setPosition($this->calculateSpacing($index));
		}
	}

	public function generateFloatingText(array $texts) {
		foreach ($texts as $index => $text) {
			$this->floatingTexts[] = new FloatingText(
				$this->calculateSpacing($index),
				$text,
				$this
			);
		}
	}

	public function all(): array {
		return $this->floatingTexts;
	}

	public function get(int $index): ?FloatingText {
		return $this->floatingTexts[$index];
	}

	public function append(FloatingText $floatingText) {
		$this->floatingTexts[] = $floatingText;
	}

	public function update(int $index, FloatingText $floatingText) {
		$this->floatingTexts[$index] = $floatingText;
	}

	public function remove(int $index) {
		unset($this->floatingTexts[$index]);
	}

	public function resetIndex() {
		$this->floatingTexts = array_values($this->floatingTexts);
	}

	public function sendToPlayer(Player $player, SendType $type) {
		foreach ($this->floatingTexts as $floatingText) {
			$floatingText->sendToPlayer($player, $type);
		}
	}

	public function sendToPlayers(array $players, SendType $type) {
		foreach ($this->floatingTexts as $floatingText) {
			$floatingText->sendToPlayers($players, $type);
		}
	}

	public function sendToLevel(Level $level, SendType $type) {
		foreach ($this->floatingTexts as $floatingText) {
			$floatingText->sendToLevel($level, $type);
		}
	}

	public function jsonSerialize(): array {
		$rounded = $this->position->round(1);
		$result = [
			Data::KEY_X => $rounded->x,
			Data::KEY_Y => $rounded->y,
			Data::KEY_Z => $rounded->z,
		];
		if (count($this->floatingTexts) >= 2) {
			$result[Data::KEY_SPACING] = [
				Data::KEY_X => $this->spacing->x,
				Data::KEY_Y => $this->spacing->y,
				Data::KEY_Z => $this->spacing->z,
			];
		}
		$this->resetIndex();
		foreach ($this->floatingTexts as $floatingText) {
			$result[Data::KEY_TEXTS][] = $floatingText->text();
		}
		return $result;
	}

	public static function fromArray(string $name, array $arr): self {
		$spacing = null;
		if (isset($arr[Data::KEY_SPACING])) {
			$spacing = new Vector3(
				$arr[Data::KEY_SPACING][Data::KEY_X],
				$arr[Data::KEY_SPACING][Data::KEY_Y],
				$arr[Data::KEY_SPACING][Data::KEY_Z]
			);
		}
		return new self(
			new Vector3(
				$arr[Data::KEY_X],
				$arr[Data::KEY_Y],
				$arr[Data::KEY_Z]
			),
			$name,
			$spacing,
			$arr[Data::KEY_TEXTS]
		);
	}

}

File: src/main/java/jp/mcbe/fuyutsuki/Texter/text/FloatingTextCluster.java
package jp.mcbe.fuyutsuki.Texter.text;

import com.google.gson.JsonObject;
import jp.mcbe.fuyutsuki.Texter.data.Data;
import jp.mcbe.fuyutsuki.Texter.util.Serializable;
import lombok.Setter;
import org.powernukkitx.Player;
import org.powernukkitx.level.Level;
import org.powernukkitx.math.Vector3;

import java.util.*;

public class FloatingTextCluster extends Nameable implements Sendable, Serializable {

	@Setter
	private Vector3 position;
	private Vector3 spacing;

	private final List<FloatingText> floatingTexts = new ArrayList<>();

	public FloatingTextCluster(Vector3 position, String name, Vector3 spacing, List<String> texts) {
		super(name);
		this.position = position;
		setSpacing(spacing);
		generateFloatingText(texts);
	}

	@Override
	public String getName() {
		return name;
	}

	public Vector3 position() {
		return position;
	}

	public Vector3 spacing() {
		return spacing;
	}

	public void setSpacing(Vector3 spacing) {
		this.spacing = spacing != null ? spacing : new Vector3(0, 0, 0);
	}

	public Vector3 calculateSpacing(int index) {
		return position.add(spacing.multiply(index));
	}

	public void recalculatePosition() {
		for (int index = 0; index < floatingTexts.size(); index++) {
			FloatingText floatingText = floatingTexts.get(index);
			if (floatingText != null) {
				floatingText.setPosition(calculateSpacing(index));
			}
		}
	}

	public void generateFloatingText(List<String> texts) {
		for (int index = 0; index < texts.size(); index++) {
			floatingTexts.add(new FloatingText(
				calculateSpacing(index),
				texts.get(index),
				this
			));
		}
	}

	public List<FloatingText> all() {
		return floatingTexts;
	}

	public FloatingText get(int index) {
		if (index < 0 || index >= floatingTexts.size()) {
			return null;
		}
		return floatingTexts.get(index);
	}

	public void append(FloatingText floatingText) {
		floatingTexts.add(floatingText);
	}

	public void update(int index, FloatingText floatingText) {
		floatingTexts.set(index, floatingText);
	}

	public void remove(int index) {
		if (index >= 0 && index < floatingTexts.size()) {
			floatingTexts.set(index, null);
		}
	}

	public void resetIndex() {
		floatingTexts.removeIf(java.util.Objects::isNull);
	}

	@Override
	public void sendToPlayer(Player player, SendType type) {
		for (FloatingText floatingText : floatingTexts) {
			if (floatingText != null) {
				floatingText.sendToPlayer(player, type);
			}
		}
	}

	@Override
	public void sendToPlayers(Collection<? extends Player> players, SendType type) {
		for (FloatingText floatingText : floatingTexts) {
			if (floatingText != null) {
				floatingText.sendToPlayers(players, type);
			}
		}
	}

	@Override
	public void sendToLevel(Level world, SendType type) {
		for (FloatingText floatingText : floatingTexts) {
			if (floatingText != null) {
				floatingText.sendToLevel(world, type);
			}
		}
	}

	@Override
	public Map<String, Object> jsonSerialize() {
		Vector3 rounded = round(position, 1);

		Map<String, Object> result = new LinkedHashMap<>();
		result.put(Data.KEY_X, rounded.x);
		result.put(Data.KEY_Y, rounded.y);
		result.put(Data.KEY_Z, rounded.z);

		if (floatingTexts.size() >= 2) {
			Map<String, Object> spacingMap = new LinkedHashMap<>();
			spacingMap.put(Data.KEY_X, spacing.x);
			spacingMap.put(Data.KEY_Y, spacing.y);
			spacingMap.put(Data.KEY_Z, spacing.z);
			result.put(Data.KEY_SPACING, spacingMap);
		}

		resetIndex();

		List<String> texts = new ArrayList<>();
		for (FloatingText floatingText : floatingTexts) {
			texts.add(floatingText.text());
		}
		result.put(Data.KEY_TEXTS, texts);

		return result;
	}

	@SuppressWarnings("unchecked")
	public static FloatingTextCluster fromMap(String name, Map<String, Object> map) {
		Vector3 spacing = null;
		if (map.containsKey(Data.KEY_SPACING)) {
			Map<String, Object> spacingMap = (Map<String, Object>) map.get(Data.KEY_SPACING);
			spacing = new Vector3(
				((Number) spacingMap.get(Data.KEY_X)).doubleValue(),
				((Number) spacingMap.get(Data.KEY_Y)).doubleValue(),
				((Number) spacingMap.get(Data.KEY_Z)).doubleValue()
			);
		}

		Vector3 position = new Vector3(
			((Number) map.get(Data.KEY_X)).doubleValue(),
			((Number) map.get(Data.KEY_Y)).doubleValue(),
			((Number) map.get(Data.KEY_Z)).doubleValue()
		);

		List<String> texts = (List<String>) map.get(Data.KEY_TEXTS);

		return new FloatingTextCluster(position, name, spacing, texts);
	}

	private static Vector3 round(Vector3 vector, int precision) {
		double factor = Math.pow(10, precision);
		return new Vector3(
			Math.round(vector.x * factor) / factor,
			Math.round(vector.y * factor) / factor,
			Math.round(vector.z * factor) / factor
		);
	}
}